### PR TITLE
[Take 2] Implement Jax-compatible PauliString

### PR DESCRIPTION
**Pull request by PhilipVinc**
This is also a remanding of the original PR by @wdphy16 , following the new infrastructure for Jax-compatible operators.

resuscitates #968

@wdphy16 as you are the original author of the code, could you please go back through it and comment it throughout? It's very complicated and it's very hard to follow what the generation of masks do...

To merge it we should document it so that in the future other people can maintain it as well.

### CODE REVIEW

**Comment by inailuig** [comment]
(_Maybe it should be mentioned that I originally ported both the ising and Paulistrings back in https://github.com/netket/netket/pull/968 to jax_) 

The paulistrings is really  just a 1:1 port of the numba code, using padding and masks wherever necessary.

**Comment by PhilipVinc** [comment]
Ah sorry, I thought it was dian.

Still, the code is very hard to follow, can you document it?

**Comment by inailuig** [comment]
> Still, the code is very hard to follow, can you document it?

It will take some time to understand again what the (also undocumented) original code was doing, then I'll  add some comments.
After re-implementing FermionOperator2nd in jax (and also porting Localoperator; faster than numba) I believe we can do slightly better than what I did here at the time, by not padding to the longest string, but instead loop over lengths.
Only benchmark will tell which approach is faster in the end.

**Comment by PhilipVinc** [comment]
Thank you. Adding some comments on the flow of the code would be great. Unfortunately, it's true that both local operator and Pauli Strings are not very documented in the deep internal of the compilation, so we could take the occasion to fix that. 

>  by not padding to the longest string, but instead loop over lengths.

If we reduce the number of strings, that's great.
My only 'requirement' is that the code must be compilable by jax so that we can use the operators in jax contexts.

**Comment by inailuig** [comment]
> Thank you. Adding some comments on the flow of the code would be great. Unfortunately, it's true that both local operator and Pauli Strings are not very documented in the deep internal of the compilation, so we could take the occasion to fix that.
> 
> > by not padding to the longest string, but instead loop over lengths.
> 
> If we reduce the number of strings, that's great. My only 'requirement' is that the code must be compilable by jax so that we can use the operators in jax contexts.

done. I dont have access to push but here is the commit: ad0c33d1de32d7a84c277873fe877b38b8afdcd6

in said commit there is a (private for now) code path which should make it competitive with IsingJax performance-wise.
I have not implemented the cutoff yet, I am not sure if we actually want/need it given that as of the current version we do the computation and use padding anyway.

**Comment by PhilipVinc** [comment]
@inailuig why don't you open the PR yourself then? after all, you are the one who wrote most of this code and I simply repackaged it a bit. I find it right that it gets attributed to you.

**Comment by PhilipVinc** [comment]
I'll also try to fix you not having the rights, but I'm in a train with terrible internet and the GitHub interface does not really work.

File: netket/operator/_pauli_strings/jax.py [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+# the general idea is the following:
+
+# observe that that Y = -i Z X
+# therefore for every pauli in a given operator we
+# (1.) apply X  (flip site)                  -- if it is X or Y
+# (2.) absorb -i to the weights              -- if it is Y
+# (3.) apply Z (pick up -1. if site is up/1) -- if it is Y or Z
+
+
+# TODO implement cutoff
```
Comment (gcarleo) [inline netket/operator/_pauli_strings/jax.py:43]: the base class supports the cutoff though? should we add an error message if somebody tries to use a cutoff with the jax version?

Comment (gcarleo) [reply netket/operator/_pauli_strings/jax.py:43]: ah ok sorry the check is already done

File: netket/operator/_pauli_strings/jax.py [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+
+        # check that it is homogeneous, throw error if it's not
+        if not isinstance(self.hilbert, HomogeneousHilbert):
+            local_states = self.hilbert.states_at_index(0)
+            if not all(
+                np.allclose(local_states, self.hilbert.states_at_index(i))
+                for i in range(self.hilbert.size)
+            ):
+                raise ValueError(
+                    "Hilbert spaces with non homogeneous local_states are not "
```
Comment (gcarleo) [inline netket/operator/_pauli_strings/jax.py:276]: does it make sense, generally speaking, to have non-homogenous hilbert spaces in this context? not sure it will be ever supported, right?

Comment (PhilipVinc) [reply netket/operator/_pauli_strings/jax.py:291]: We could eventually support it but it does not really make sense, indeed.
